File: tools/axiom/axiom_jumplist_parser.py
import os
import pandas as pd
from utils.discovery import find_artifact_files, load_csv_with_progress
from collector.collector import add_rows
from utils.logger import print_and_log
from utils.summary import track_summary

def process_axiom_jumplist(axiom_dir: str, batch_size: int, base_dir: str):
    artifact_name = "Axiom_JumpLists"
    print_and_log(f"[{artifact_name}] Scanning for relevant CSVs under: {axiom_dir}")

    jump_files = find_artifact_files(axiom_dir, base_dir, artifact_name)
    if not jump_files:
        print_and_log(f"[{artifact_name}] No JumpList files found.")
        return

    for file_path in jump_files:
        print_and_log(f"[{artifact_name}] Processing: {file_path}")
        total_rows = 0
        try:
            for df in load_csv_with_progress(file_path, batch_size, artifact_name=artifact_name):
                timeline_data = []
                date_columns = [
                    ("Target File Created Date/Time - UTC+00:00 (M/d/yyyy)", "Target Created"),
                    ("Target File Last Modified Date/Time - UTC+00:00 (M/d/yyyy)", "Target Modified"),
                    ("Target File Last Accessed Date/Time - UTC+00:00 (M/d/yyyy)", "Target Accessed"),
                    ("Last Access Date/Time - UTC+00:00 (M/d/yyyy)", "Source Accessed")
                ]

                for _, row in df.iterrows():
                    for time_field, label in date_columns:
                        timestamp = row.get(time_field, "")
                        dt = pd.to_datetime(timestamp, utc=True, errors="coerce")
                        if pd.isnull(dt):
                            continue

                        target = next(
                            (
                                val for val in [
                                    row.get("Linked Path"),
                                    row.get("Location"),
                                    row.get("Source")
                                ] if isinstance(val, str) and val.strip()
                            ),
                            ""
                        )

                        if target:
                            if os.path.splitext(target)[1]:
                                details = os.path.basename(target)
                            else:
                                details = os.path.basename(os.path.normpath(target))
                        else:
                            details = row.get("Potential App Name", "")

                        timeline_row = {
                            "DateTime": dt.isoformat().replace("+00:00", "Z"),
                            "Tool": "Axiom",
                            "ArtifactName": "JumpLists",
                            "TimestampInfo": label,
                            "Description": "File & Folder Access",
                            "DataPath": target,
                            "DataDetails": details,
                            "FileSize": row.get("Target File Size (Bytes)", ""),
                            "EvidencePath": os.path.relpath(file_path, base_dir) if base_dir else file_path
                        }

                        timeline_data.append(timeline_row)

                total_rows += len(timeline_data)
                add_rows(timeline_data)

            track_summary("Axiom", artifact_name, total_rows)
            print_and_log(f"[✓] Parsed {total_rows} timeline rows from: {file_path}")

        except Exception as e:
            print_and_log(f"[{artifact_name}] Failed to parse {file_path}: {e}")
            continue

File: ForensicTimeliner.Core/Tools/Axiom/AxiomJumplistParser.cs
using CsvHelper;
using CsvHelper.Configuration;
using ForensicTimeliner.Models;
using ForensicTimeliner.Interfaces;
using ForensicTimeliner.Models;
using ForensicTimeliner.Utils;
using Spectre.Console;
using System.Globalization;

namespace ForensicTimeliner.Tools.Axiom;

public class AxiomJumpListsParser : IArtifactParser
{
    public List<TimelineRow> Parse(string inputDir, string baseDir, ArtifactDefinition artifact, ParsedArgs args)
    {
        var rows = new List<TimelineRow>();

        Logger.PrintAndLog($"[>] - [{artifact.Artifact}] Scanning for relevant CSVs under: [{inputDir}]", "SCAN");

        var files = Discovery.FindArtifactFiles(inputDir, baseDir, artifact.Artifact);
        if (!files.Any())
        {
            Logger.PrintAndLog($"[!] - [{artifact.Artifact}] No matching files found in: {inputDir}", "WARN");
            return rows;
        }

        foreach (var file in files)
        {
            int timelineCount = 0;
            Logger.PrintAndLog($"[+] - [{artifact.Artifact}] Processing: {Path.GetRelativePath(baseDir, file)}", "PROCESS");

            try
            {
                using var reader = new StreamReader(file);
                using var csv = new CsvReader(reader, new CsvConfiguration(CultureInfo.InvariantCulture)
                {
                    HeaderValidated = null,
                    MissingFieldFound = null
                });

                var records = csv.GetRecords<dynamic>();
                foreach (var record in records)
                {
                    var dict = (IDictionary<string, object>)record;

                    var dateColumns = new Dictionary<string, string>
                    {
                        { "Target File Created Date/Time - UTC+00:00 (M/d/yyyy)", "Target Created" },
                        { "Target File Last Modified Date/Time - UTC+00:00 (M/d/yyyy)", "Target Modified" },
                        { "Target File Last Accessed Date/Time - UTC+00:00 (M/d/yyyy)", "Target Accessed" },
                        { "Last Access Date/Time - UTC+00:00 (M/d/yyyy)", "Source Accessed" }
                    };

                    foreach (var field in dateColumns)
                    {
                        var parsedDt = dict.GetDateTime(field.Key);
                        if (parsedDt == null) continue;

                        string dtStr = parsedDt.Value.ToString("o").Replace("+00:00", "Z");

                        // Determine best available DataPath
                        string target = dict.GetString("Linked Path") ??
                                        dict.GetString("Location") ??
                                        dict.GetString("Source") ?? "";

                        string details;
                        if (!string.IsNullOrWhiteSpace(target))
                        {
                            if (Path.HasExtension(target))
                            {
                                details = Path.GetFileName(target);
                            }
                            else
                            {
                                details = Path.GetFileName(Path.GetDirectoryName(target));
                            }
                        }
                        else
                        {
                            details = dict.GetString("Potential App Name");
                        }

                        rows.Add(new TimelineRow
                        {
                            DateTime = dtStr,
                            TimestampInfo = field.Value,
                            ArtifactName = "JumpLists",
                            Tool = artifact.Tool,
                            Description = "File & Folder Access",
                            DataPath = target,
                            DataDetails = details,
                            FileSize = dict.GetLong("Target File Size (Bytes)"),
                            EvidencePath = Path.GetRelativePath(baseDir, file)
                        });

                        timelineCount++;
                    }
                }

                Logger.PrintAndLog($"[✓] - [{artifact.Artifact}] Parsed {timelineCount} timeline rows from: {Path.GetFileName(file)}", "SUCCESS");
                LoggerSummary.TrackSummary(artifact.Tool, artifact.Artifact, timelineCount);
            }
            catch (Exception ex)
            {
                Logger.PrintAndLog($"[{artifact.Artifact}] Failed to parse {file}: {ex.Message}", "ERROR");
            }
        }

        return rows;
    }
}


